### ISSUE TITLE
Training on new datasets

### ISSUE THREAD

**Issue by raphaeli** [vlfeat/matconvnet#59]
Hi,
This is not an issue question but I will ask if someone has applied the code on datasets other than the ones of the examples. Say:
https://www.kaggle.com/c/datasciencebowl/data

Although tried hard and the code runs OK the results are poor. If someone would comment on this or on the net below I would be greatful

function cnn_plankton(varargin)

run(fullfile(fileparts(mfilename('fullpath')), ...
  '..', 'matlab', 'vl_setupnn.m')) ;

opts.dataDir = fullfile('data','plankton') ;
opts.expDir = fullfile('data','plankton-baseline') ;
opts.imdbPath = fullfile(opts.expDir, 'imdb.mat');
opts.train.batchSize = 256 ;
opts.train.numEpochs = 100 ;
opts.train.continue = true ;
opts.train.useGpu = true ;
opts.train.learningRate = [0.001*ones(1, 12) 0.0001*ones(1,6) 0.00001] ;
opts.train.expDir = opts.expDir ;
opts = vl_argparse(opts, varargin) ;
 
% --------------------------------------------------------------------
%                                                         Prepare data
% --------------------------------------------------------------------

if exist(opts.imdbPath, 'file')
  imdb = load(opts.imdbPath) ;
else
  imdb = getPlanktonImdb(opts) ;
  mkdir(opts.expDir) ;
  save(opts.imdbPath, '-struct', 'imdb') ;
end

f=1e-2;

net.layers = {} ;
% 1 conv1
net.layers{end+1} = struct('type', 'conv', ...
'filters', f*randn(5,5,1,32, 'single'), ...
'biases', zeros(1, 32, 'single'), ...
'stride', 2, ...
'pad', 2) ;

%net.layers{end+1} = struct('type', 'normalize', ...
%                           'param', [1 1 0.0001/5 0.95]) ;

net.layers{end+1} = struct('type', 'relu') ;
net.layers{end+1} = struct('type', 'pool', ...
                           'method', 'max', ...
                           'pool', [3 3], ...
                           'stride', 2, ...
                           'pad', 0) ;

net.layers{end+1} = struct('type', 'conv', ...
'filters', f*randn(3,3,32,64, 'single'),...
'biases', zeros(1,64,'single'), ...
'stride', 1, ...
'pad', 0) ;
% 5 relu2
net.layers{end+1} = struct('type', 'relu') ;
% 6 pool2 (avg pool)
net.layers{end+1} = struct('type', 'conv', ...
'filters', 0.01*randn(3,3,64,128, 'single'),...
'biases', zeros(1,128,'single'), ...
'stride', 1, ...
'pad', 1) ;
% 8 relu3
net.layers{end+1} = struct('type', 'relu') ;
% 9 pool3 (avg pool)
net.layers{end+1} = struct('type', 'pool', ...
                           'method', 'max', ...
                           'pool', [3 3], ...
                           'stride', 1, ...
                           'pad', 0) ;
% 10 ip1
net.layers{end+1} = struct('type', 'conv', ...
'filters', f*randn(1,1,128,512, 'single'),...
'biases', zeros(1,512,'single'), ...
'stride', 1, ...
'pad', 0) ;
net.layers{end+1} = struct('type', 'relu') ;
%net.layers{end+1} = struct('type', 'dropout', ...
%                           'rate', 0.5) ;
net.layers{end+1} = struct('type', 'conv', ...
'filters', f*randn(1,1,512,121, 'single'),...
'biases', zeros(1,121,'single'), ...
'stride', 1, ...
'pad', 0) ;
% 12 loss 
net.layers{end+1} = struct('type', 'softmaxloss') ;
% --------------------------------------------------------------------
%                                                                Train
% --------------------------------------------------------------------

% Take the mean out and make GPU if needed

%imdb.images.data = bsxfun(@minus, imdb.images.data, mean(imdb.images.data,4)) ;
if opts.train.useGpu
  imdb.images.data = gpuArray(imdb.images.data) ;
end

[net,info] = cnn_train(net, imdb, @getBatch, ...
    opts.train, ...
    'val', find(imdb.images.set == 3)) ;

% --------------------------------------------------------------------
function [im, labels] = getBatch(imdb, batch)
% --------------------------------------------------------------------
ind = randi(7,1,numel(batch));
im = imdb.images.data(:,:,:,batch) ;
im = circshift(im, [randi([-1,1]),randi([-1,1])]);
im(:,:,:,find(ind == 2)) = flipdim(im(:,:,:,find(ind == 2)),2);
im(:,:,:,find(ind == 3)) = imrotate(im(:,:,:,find(ind == 3)),45,'bilinear','crop');
im(:,:,:,find(ind == 4)) = imrotate(im(:,:,:,find(ind == 4)),-45,'bilinear','crop');
im(:,:,:,find(ind == 5)) = imrotate(im(:,:,:,find(ind == 5)),90,'bilinear','crop');
im(:,:,:,find(ind == 6)) = imrotate(im(:,:,:,find(ind == 6)),-90,'bilinear','crop');
im(:,:,:,find(ind == 7)) = flipdim(im(:,:,:,find(ind == 7)),1);

labels = imdb.images.labels(1,batch) ;

% --------------------------------------------------------------------
function imdb = getPlanktonImdb(opts)
% --------------------------------------------------------------------
clNames = 'acantharia_protist_big_center,acantharia_protist_halo,acantharia_protist,amphipods,appendicularian_fritillaridae,appendicularian_s_shape,appendicularian_slight_curve,appendicularian_straight,artifacts_edge,artifacts,chaetognath_non_sagitta,chaetognath_other,chaetognath_sagitta,chordate_type1,copepod_calanoid_eggs,copepod_calanoid_eucalanus,copepod_calanoid_flatheads,copepod_calanoid_frillyAntennae,copepod_calanoid_large_side_antennatucked,copepod_calanoid_large,copepod_calanoid_octomoms,copepod_calanoid_small_longantennae,copepod_calanoid,copepod_cyclopoid_copilia,copepod_cyclopoid_oithona_eggs,copepod_cyclopoid_oithona,copepod_other,crustacean_other,ctenophore_cestid,ctenophore_cydippid_no_tentacles,ctenophore_cydippid_tentacles,ctenophore_lobate,decapods,detritus_blob,detritus_filamentous,detritus_other,diatom_chain_string,diatom_chain_tube,echinoderm_larva_pluteus_brittlestar,echinoderm_larva_pluteus_early,echinoderm_larva_pluteus_typeC,echinoderm_larva_pluteus_urchin,echinoderm_larva_seastar_bipinnaria,echinoderm_larva_seastar_brachiolaria,echinoderm_seacucumber_auricularia_larva,echinopluteus,ephyra,euphausiids_young,euphausiids,fecal_pellet,fish_larvae_deep_body,fish_larvae_leptocephali,fish_larvae_medium_body,fish_larvae_myctophids,fish_larvae_thin_body,fish_larvae_very_thin_body,heteropod,hydromedusae_aglaura,hydromedusae_bell_and_tentacles,hydromedusae_h15,hydromedusae_haliscera_small_sideview,hydromedusae_haliscera,hydromedusae_liriope,hydromedusae_narco_dark,hydromedusae_narco_young,hydromedusae_narcomedusae,hydromedusae_other,hydromedusae_partial_dark,hydromedusae_shapeA_sideview_small,hydromedusae_shapeA,hydromedusae_shapeB,hydromedusae_sideview_big,hydromedusae_solmaris,hydromedusae_solmundella,hydromedusae_typeD_bell_and_tentacles,hydromedusae_typeD,hydromedusae_typeE,hydromedusae_typeF,invertebrate_larvae_other_A,invertebrate_larvae_other_B,jellies_tentacles,polychaete,protist_dark_center,protist_fuzzy_olive,protist_noctiluca,protist_other,protist_star,pteropod_butterfly,pteropod_theco_dev_seq,pteropod_triangle,radiolarian_chain,radiolarian_colony,shrimp_caridean,shrimp_sergestidae,shrimp_zoea,shrimp-like_other,siphonophore_calycophoran_abylidae,siphonophore_calycophoran_rocketship_adult,siphonophore_calycophoran_rocketship_young,siphonophore_calycophoran_sphaeronectes_stem,siphonophore_calycophoran_sphaeronectes_young,siphonophore_calycophoran_sphaeronectes,siphonophore_other_parts,siphonophore_partial,siphonophore_physonect_young,siphonophore_physonect,stomatopod,tornaria_acorn_worm_larvae,trichodesmium_bowtie,trichodesmium_multiple,trichodesmium_puff,trichodesmium_tuft,trochophore_larvae,tunicate_doliolid_nurse,tunicate_doliolid,tunicate_partial,tunicate_salp_chains,tunicate_salp,unknown_blobs_and_smudges,unknown_sticks,unknown_unclassified';
clNames = strsplit(clNames,',');
c = 1;
maxpix = 48;
X1 = zeros(maxpix,maxpix,1,30336);
labels = zeros(1,30336);
for i=1:length(clNames),
    y2 = dir(fullfile(['/home/midas/matconvnet/examples/' opts.dataDir], 'train','/',clNames{i}));
    
    for j=3:length(y2),
        fname = fullfile(['/home/midas/matconvnet/examples/' opts.dataDir], 'train','/',clNames{i},'/',y2(j).name);
        im = imread(fname) ;
        im_ = 255-imresize(im, [maxpix,maxpix]);
        X1(:,:,1,c) = im_;
        labels(c) = i;
        c=c+1;
    end
end

X1 = bsxfun(@minus, X1, mean(X1, 4));
X1 = single(X1);
labels = double(labels);
imdb.images.data = X1 ;
imdb.images.labels = labels ;
imdb.images.set = [ones(1,29000) 3*ones(1,1336)] ;
imdb.meta.sets = {'train', 'val', 'test'} ;
imdb.meta.classes = clNames;
